# Introduction

The [HHL algorithm](https://en.wikipedia.org/wiki/Quantum_algorithm_for_linear_systems_of_equations) underlies many quantum machine learning protocols, but it is a **highly nontrivial algorithm with lots of conditions**. In this notebook, we implement the algorithm to gain a better understanding of how it works and when it works efficiently.

(*The notebook is derived from the [computational appendix](https://gitlab.com/apozas/bayesian-dl-quantum) of the paper [Bayesian Deep Learning on a Quantum Computer](https://arxiv.org/abs/1806.11463). We restrict our attention to inverting a $2\times 2$ matrix, following Pan et al's implementation [[1](#1)] of the algorithm.*)

In [2]:
import numpy as np
from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister, execute, BasicAer

π = np.pi

# Setting up the problem

The objective is to solve the linear equation $Ax=b$ with $A = \frac{1}{2}\begin{bmatrix}3 & 1 \\1 & 3 \\ \end{bmatrix}$ and $b =\begin{bmatrix} 1 \\ 0 \\ \end{bmatrix}$. 

### Classical solution

In order to be prepared for the following, let us start by solving this equation using "traditional" techniques...

In [3]:
A = 0.5 * np.array([[3, 1], [1, 3]])
b = np.array([1, 0])

x = np.linalg.solve(A, b)
x /= np.linalg.norm(x)

print('(normalized) solution is %s' % x)

(normalized) solution is [ 0.9486833  -0.31622777]


We see that the expected solution to this linear equation is $\approx 0.949 |0 \rangle - 0.316 |1\rangle$.

### Solving using HHL on a quantum computer

We will encode the $A$ matrix as a Hamiltonian and $b$ in a register. With ancillas, we will need a total of four qubits and one classical register for post-selection. We add an extra qubit and extra classical register to create a swap test to compare our result to the ideal state.

In [4]:
q = QuantumRegister(6)
c = ClassicalRegister(2)
hhl = QuantumCircuit(q, c)

The vector $b$ can be (amplitude) encoded as $\left|b\right\rangle = \sum_{i=0}^N b_i\left|i\right\rangle = \left|0\right\rangle$, so no explicit state preparation circuit is needed for this case (this will not be true in general).

# Eigenvalue (phase) estimation

The next step is to encode the eigenvalues of the matrix $A$ in an additional register. This is done via quantum phase estimation of the evolution described by the Hamiltonian $A$ during some time $t_0$, $\exp(i A t_0)$. The protocol has three steps.

- First we prepare the ancilla state $\left|\psi_0\right\rangle=\sum_{\tau=0}^{T-1}\left|\tau\right\rangle$. Why this state? It will control the time evolution: it is **like a clock turning on evolution for a certain amount of time**. The original HHL algorithm suggests a weighted superposition of all states $\tau$ that minimizes errors in following steps in the algorithm. However, for our implementation, a uniform superposition already gives good results.


- Our goal is to **create a superposition of $A$ as a Hamiltonian applied for different durations**. Since the eigenvalues are always situated on the complex unit circle, these differently evolved components in the superposition help reveal the eigenstructure. So we apply the conditional Hamiltonian evolution

$$\sum_{\tau=0}^{T-1}\left|\tau\right\rangle\left\langle\tau\right|\otimes e^{i A\tau t_0/T} \quad \text{on} \quad \left|\psi_0\right\rangle\otimes\left|b\right\rangle$$

**This operation evolves the state $\left|b\right\rangle$ according to the Hamiltonian $A$ for the time $\tau$ determined by the state $\left|\psi_0\right\rangle$**. Given that in $\left|\psi_0\right\rangle$ we have a superposition of all possible time steps between $0$ and $T$, we will end up with a superposition of all possible evolutions, and a suitable choice of number of timesteps $T$ and total evolution time $t_0$ allow to encode binary representations of the eigenvalues.


- As a final step, we apply an inverse Fourier transformation that writes the phases (that, recall, encode the eigenvalues of $A$) into new registers.

The total circuit for phase estimation was shown in the previous notebook.

In our $2\times 2$ case, the circuit is massively simplified. **Cheating a little more**: Given that the matrix $A$ has eigenvalues that are powers of $2$, we can choose $T=4$ and $t_0=2\pi$ to obtain exact results with just two controlled evolutions.

In [5]:
# Superposition
hhl.h(q[1])
hhl.h(q[2])

# Controlled-U0
hhl.cu3(-π / 2, -π / 2, π / 2, q[2], q[3])
hhl.cu1(3 * π / 4, q[2], q[3])
hhl.cx(q[2], q[3])
hhl.cu1(3 * π / 4, q[2], q[3])
hhl.cx(q[2], q[3])

# Controlled-U1
hhl.cx(q[1], q[3]);

We apply quantum inverse Fourier transformation to write the phase to a register:

In [6]:
hhl.swap(q[1], q[2])
hhl.h(q[2])
hhl.cu1(-π / 2, q[1], q[2])
hhl.h(q[1]);

The state of the system after this decomposition is approximately $\sum _{j{\mathop {=}}1}^{N}\beta _{j}\left|u_{j}\right\rangle \left|\lambda_{j}\right\rangle$, where $\left|b\right\rangle=\sum _{j{\mathop {=}}1}^{N}\beta _{j}\left|u_{j}\right\rangle$ is the encoding of the vector $b$ in the eigenbasis of $A$. 

Now, there is an often overlooked step that performs bit operations on $\left|\lambda_{j}\right\rangle$ to actually invert it.

**Is it another cheat?** In our case, the inversion of the eigenvalues is easy. The eigenvalues of $A$ are $\lambda_1=2=10_2$ and $\lambda_2=1=01_2$, and their reciprocals are $\lambda_1^{-1}=1/2$ and $\lambda_2^{-1}=1$. Noting that $2\lambda_1^{-1}=01_2$ and $2\lambda_2^{-1}=10_2$, a swap gate is enough to obtain the state $\sum _{j{\mathop {=}}1}^{N}\beta _{j}\left|u_{j}\right\rangle \left|2\lambda _{j}^{-1}\right\rangle$, that encodes the reciprocals of the eigenvalues.

In [7]:
hhl.swap(q[1], q[2]);

# Conditional rotation of ancilla

Next, we perform a conditional rotation to encode the information of the reciprocals of the eigenvalues in the amplitudes of a state, on which we will later post-select. The state we would like to get is 

$$\sum _{j{\mathop {=}}1}^{N}\beta _{j}\left|u_{j}\right\rangle\left|2\lambda _{j}^{-1}\right\rangle \left(\sqrt{1-\frac{C^2}{\lambda_j^2}}\left|0\right\rangle+\frac{C}{\lambda_j}\left|1\right\rangle \right)$$

This is achieved by controlled rotations in the same spirit of the conditional Hamiltonian evolution.

In [8]:
hhl.cu3(0.392699, 0, 0, q[1], q[0])  # Controlled-RY0
hhl.cu3(0.19634955, 0, 0, q[2], q[0]);  # Controlled-RY1

*Where these numbers come has been the source of questions in the online comments...*

# Uncomputing the eigenvalue register

A necessary step when performing quantum computations is to uncompute all operations except those that store the information that we want to obtain from the algorithm in the final registers. We need to do this in case the registers are entangled, which would affect the results.

In our case, we must uncompute the phase estimation protocol. After the uncomputation, the state should be

$$\sum_{j=1}^N\beta_j\left|u_j\right\rangle\left|0\right\rangle\left(\sqrt{1-\frac{C^2}{\lambda_j^2}}\left|0\right\rangle+\frac{C}{\lambda_j}\left|1\right\rangle \right)$$

so we can safely forget about the eigenvalue register.

In [9]:
hhl.swap(q[1], q[2])
hhl.h(q[1])
hhl.cu1(π / 2, q[1], q[2])  # Inverse(Dagger(Controlled-S))
hhl.h(q[2])
hhl.swap(q[2], q[1])

# Inverse(Controlled-U1)
hhl.cx(q[1], q[3])

# Inverse(Controlled-U0)
hhl.cx(q[2], q[3])
hhl.cu1(-3 * π / 4, q[2], q[3])
hhl.cx(q[2], q[3])
hhl.cu1(-3 * π / 4, q[2], q[3])
hhl.cu3(-π / 2, π / 2, -π / 2, q[2], q[3])

# End of Inverse(Controlled-U0)
hhl.h(q[2])
hhl.h(q[1]);

# Rejection sampling on the ancilla register and a swap test

The state

$$\left|x\right\rangle=A^{-1}\left|b\right\rangle\propto\sum_j \beta_j\lambda_j^{-1}\left|u_j\right\rangle$$

that contains information about the solution to $Ax=b$ is that obtained when measuring $|1\rangle$ on the ancilla state. We perform the post-selection by projecting onto the desired $\left|1\right\rangle$. To check that the solution is the expected one, we prepare the correct output state manually to perform a swap test with the outcome.

In [10]:
# Target state preparation
hhl.rz(-π, q[4])
hhl.u1(π, q[4])
hhl.h(q[4])
hhl.ry(-0.9311623288419387, q[4])
hhl.rz(π, q[4]);

It is a good exercise to check that the right result is given by the state $\left|x\right\rangle=0.949\left|0\right\rangle + 0.314\left|1\right\rangle$, which is the state we prepare above.

In [11]:
def verifExpectedState():
    q, c = QuantumRegister(1), ClassicalRegister(1)
    circuit = QuantumCircuit(q, c)
    circuit.rz(-π, q[0])
    circuit.u1(π, q[0])
    circuit.h(q[0])
    circuit.ry(-0.9311623288419387, q[0])
    circuit.rz(π, q[0])

    backend_statevector = BasicAer.get_backend('statevector_simulator')
    job = execute(circuit, backend_statevector)
    return job.result().get_statevector(circuit)

verifExpectedState()

array([ 0.94929297+0.j, -0.31439284+0.j])

This confirms that the target state with which we are going to compare the output of the ``hhl`` circuit is indeed $\approx 0.949 |0\rangle - 0.314|1\rangle$ in accordance with what we had previously found in the classical solution at the beginning of the notebook.

Next, using this prepared state that we know to be the exact solution, we compare it to the outcome of our ``hhl`` circuit using a ``swap test``.   **Swap tests can be used to test two unknown pure states for equality**.  Given two states $|\psi \rangle$ and $|\psi^\prime \rangle$ a swap test circuit returns the overlap $|\langle \psi | \psi^\prime \rangle|^2$.



In other words, we **cannot extract the solution from the quantum circuit but only verify** (assuming that we have already solved the problem some other way) that the state is indeed the expected one.

In [12]:
# Swap test
hhl.h(q[5])
hhl.cx(q[4], q[3])
hhl.ccx(q[5], q[3], q[4])
hhl.cx(q[4], q[3])
hhl.h(q[5])

# Final measurements
hhl.barrier(q)
hhl.measure(q[0], c[0])
hhl.measure(q[5], c[1]);

There are two measurements performed, one of the ancilla register (for doing the post-selection) and another one that gives the result of the swap test.

In [13]:
def get_psuccess(counts):
    succ_rotation_fail_swap = counts['11'] if '11' in counts.keys() else 0
    succ_rotation_succ_swap = counts['01']
    
    succ_rotation = succ_rotation_succ_swap + succ_rotation_fail_swap
    return succ_rotation_succ_swap / succ_rotation

Finally we run the circuit on the simulator:

In [14]:
backend = BasicAer.get_backend('qasm_simulator')
job = execute(hhl, backend, shots=1000)
result = job.result()
counts = result.get_counts(hhl)

counts

{'00': 985, '01': 15}

This concludes our demonstration that the ``hhl`` circuit designed above indeed revovers the expected solution to the linear equation.

In [15]:
get_psuccess(counts)

1.0